[MRG] Add verbose flag and callback to *_minimize

MechCoder opened the pull request:
Fixes https://github.com/scikit-optimize/scikit-optimize/issues/111

MechCoder commented:
I added some basic information about the function evaluation. What else do we need?

@betatim @glouppe

glouppe commented on skopt/dummy_opt.py:
```diff
@@ -34,6 +35,10 @@ def dummy_minimize(func, dimensions, n_calls=100, random_state=None):
         Set random state to something other than None for reproducible
         results.
 
+    * `verbose` [int, default=False]:
```
`int` -> `boolean`?

glouppe commented:
Thanks! Not sure if it is an overkill, but what do you think of a joblib-like output? How the verbose mode is structured in https://github.com/joblib/joblib/blob/62c0914fb52e69360c8724183f6d06c1e236532b/joblib/parallel.py is also worth looking at

nfcampos commented:
how about total elapsed time and iteration time?

MechCoder commented:
I'll finish this up after https://github.com/scikit-optimize/scikit-optimize/pull/119 has been merged.

glouppe commented:
Regarding the code structure, we could also have an optional `callback` argument, called at the end of each iteration and through which the current `OptimizeResult` object is passed. Logging could then be implemented as one such callback.

MechCoder commented:
OK, I've added the time taken and a callback argument that takes the res as an argument.

I feel a more useful verbose flag can be added after we have refactored the minimize code later.

glouppe commented on skopt/utils.py:
```diff
@@ -0,0 +1,20 @@
+import numpy as np
+from scipy.optimize import OptimizeResult
+
+def pack_optimize_result(Xi, yi, space, rng, specs, models=None):
```
Shall we make all arguments optional to make this function more versatile?

glouppe replied:
Shall we aim for a shorter function name? e.g. simply `pack`? (which should be clear from the context)

glouppe commented on skopt/utils.py:
```diff
@@ -0,0 +1,20 @@
+def pack_optimize_result(Xi, yi, space, rng, specs, models=None):
+    """
+    Instantiate OptimizeResult() with the required information.
+    """
```
Please also document the argument

glouppe commented on skopt/tree_opt.py:
optional

glouppe commented:
A couple of comments:
- `callback` should accept either a single function or a list of functions.
- verbose mode should be implemented as a callback added to the list of callbacks -- this would factor out all the if-statements repeated in all optimizers.

MechCoder commented:
Sorry, I do not understand, I think callback should just accept `res` and whatever the user wants should be called internally in callback. Right now, it is quite similar to the callback in http://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.fmin_l_bfgs_b.html#scipy.optimize.fmin_l_bfgs_b . What are the use case you are looking at?

(If you want to refactor the verbose calls in a private function, I can do that)

MechCoder commented:
I refactored the verbose calls to a function in `skopt.utils`

betatim commented on skopt/tree_opt.py:
```diff
@@ -265,7 +265,7 @@ def gbrt_minimize(func, dimensions, base_estimator=None, n_calls=100,
 
-    * `callback` [callable, optiona]
+    * `callback` [callable, optional]
         If provided, then `callback(res)` is called after call to func.
```
"after call" -> "after each call to `func`"

MechCoder replied:
the trouble with copy pasting this stuff, is that I need to copy paste it everywhere again ;)

betatim commented on skopt/utils.py:
```diff
@@ -1,9 +1,35 @@
 from scipy.optimize import OptimizeResult
 
-def pack_optimize_result(Xi, yi, space, rng, specs, models=None):
+def set_results(Xi, yi, space=None, rng=None, specs=None, models=None):
```
at first reading (with first ☕ of the day) `set_results` made me think "ah, they set some properties on the first argument". Would `create_result` be a better name?

betatim commented:
I think @glouppe meant that the `*_minimize` argument named `callback` should accept either a single callable or a list of them.

MechCoder commented:
Ah I see, that makes sense. I was misinterpreting it as the function provided as `callback` should accept a callable which seemed odd to me :P

betatim commented:
My two comments where on an old commit it seems :-( mainly about language tuning.

MechCoder commented:
I've addressed those and made callback accept a list of callables.

betatim commented on skopt/dummy_opt.py:
```diff
@@ -4,16 +4,19 @@
 import inspect
 import numbers
 import numpy as np
+from time import time
```
`from collections import Callable` missing. Can you also make a common test to exercise the API (checking that an exception is raised if `callable` isn't a list of callables or a callable).

betatim commented on skopt/dummy_opt.py:
```diff
@@ -4,16 +4,19 @@
 from .space import Space
-
+from .utils import create_result
+from .utils import verbose_func
```
Should have commented on this before: can we think up a better name? I can't suggest one right now but `verbose_func` seems overly generic and "I won't tell you what I actually do"

MechCoder replied:
It's difficult to think of one. There is quite a bit information to be represented in a single name :( (minimum, time taken and function value obtained).

MechCoder replied:
We can have a naming improvement issue, where we resolve all such issues just before the release. (including the project name ;) )

MechCoder replied:
Maybe `progess_func`? Idk

betatim replied:
Tech debt accrued now will have to be repaid with interest later.

betatim commented on skopt/gp_opt.py:
```diff
@@ -209,8 +230,21 @@ def gp_minimize(func, dimensions, base_estimator=None, alpha=10e-10,
     if y0 is None and x0:
-        y0 = [func(x) for x in x0]
+        y0 = []
+        for i, x in enumerate(x0):
```
is `i` used anywhere? 🔍

MechCoder replied:
it is used below in the result instantiation.

MechCoder replied:
Oh, sorry I was looking at a different `for` block

MechCoder replied:
oops, seems like you caught a bug.

betatim commented on skopt/gp_opt.py:
```diff
@@ -174,6 +187,14 @@ def gp_minimize(func, dimensions, base_estimator=None, alpha=10e-10,
     specs = {"args": copy.copy(inspect.currentframe().f_locals),
              "function": inspect.currentframe().f_code.co_name}
 
+    if callback is not None:
```
Should we start the refactoring campaign for repeated stuff across the `*_minimize` functions in this PR?

This bit looks like an ideal candidate for: `callback = check_callback(callback)` where `check_callback` lives in `skopt.utils` or some such (inspired by `check_array` and friends in `sklearn`).

betatim commented on skopt/gp_opt.py:
```diff
@@ -209,8 +230,21 @@ def gp_minimize(func, dimensions, base_estimator=None, alpha=10e-10,
-        y0 = [func(x) for x in x0]
+        y0 = []
+        for i, x in enumerate(x0):
+
```
👻 why the white space?

betatim commented on skopt/gp_opt.py:
```diff
@@ -227,9 +261,25 @@ def gp_minimize(func, dimensions, base_estimator=None, alpha=10e-10,
+    yi = y0
+
+    for i, x in enumerate(X_rand):
+
```
👻 why the white space?

betatim commented on skopt/gp_opt.py:
```diff
@@ -246,6 +296,9 @@ def gp_minimize(func, dimensions, base_estimator=None, alpha=10e-10,
     models = []
     n_model_iter = n_calls - n_total_init_calls
     for i in range(n_model_iter):
+
```
why the white face (face <-> space, get it)? Did you see a 👻 ? (sorry, spending too much time in the UK with people making bad puns)

MechCoder replied:
Do you have a github emoji guide or is there a dropdown that I am missing somewhere?

betatim replied:
?? You mean how do I know what emoji exist? The answer to that is: too much time on slack.

betatim commented on skopt/tests/test_common.py:
```diff
@@ -64,25 +64,28 @@ def check_minimizer_bounds(result):
+        yield (check_minimizer_api, result, 0)
         yield (check_minimizer_bounds, result)
-        assert_raises(ValueError, minimizer, lambda x: x, [[-5, 10]])
+        assert_raises(ValueError, dummy_minimize, lambda x: x, [[-5, 10]])
```
Do you remember which `ValueError` this is? Can you add a regex match to this (or create an issue reminding us to create a regex match)

MechCoder replied:
I think it is that the function should return a scalar.

betatim commented on skopt/tree_opt.py:
```diff
@@ -4,6 +4,7 @@
 import inspect
 import numbers
 import numpy as np
```
import Callable missing

betatim commented on skopt/tree_opt.py:
```diff
@@ -19,11 +20,23 @@
                    n_points, n_random_starts, x0=None, y0=None,
-                   random_state=None, acq="EI", xi=0.01, kappa=1.96):
+                   random_state=None, acq="EI", xi=0.01, kappa=1.96,
+                   verbose=False, specs=None, callback=None):
```
Is it legal for `specs` to ever not be passed as argument? I think it should be a mandatory argument.

betatim commented on skopt/tree_opt.py:
```diff
@@ -53,8 +66,20 @@ def _tree_minimize(func, dimensions, base_estimator, n_calls,
     if y0 is None and x0:
-        y0 = [func(x) for x in x0]
+        y0 = []
+        for i, x in enumerate(x0):
```
`i`? 🔍

betatim commented on skopt/tree_opt.py:
```diff
@@ -53,8 +66,20 @@ def _tree_minimize(func, dimensions, base_estimator, n_calls,
-        y0 = [func(x) for x in x0]
+        y0 = []
+        for i, x in enumerate(x0):
+
```
👻

betatim commented on skopt/tree_opt.py:
```diff
@@ -70,15 +95,35 @@ def _tree_minimize(func, dimensions, base_estimator, n_calls,
+    yi = y0
+
+    for i, x in enumerate(X_rand):
+
```
👻

betatim commented on skopt/tree_opt.py:
```diff
@@ -70,15 +95,35 @@ def _tree_minimize(func, dimensions, base_estimator, n_calls,
     models = []
     n_model_iter = n_calls - n_total_init_calls
     for i in range(n_model_iter):
+
```
👻

betatim commented on skopt/utils.py:
```diff
@@ -0,0 +1,87 @@
+from scipy.optimize import OptimizeResult
+
+
+def create_result(Xi, yi, space=None, rng=None, specs=None, models=None):
```
should we represent the absence of models by `None` or an empty list in `OptimizeResult`? I've seen `[]` somewhere else in the code but I think `None` is a better choice.

What ever we decide, the choice should be checked for in `check_minimizer_api` int `test_common.py`

MechCoder replied:
I agree that None is a better option. I fixed it.

betatim commented:
Looks like a lot of comments but mostly it is repeats. Biggest point is: can we come up with a better name for `verbose_func`?

There is a missing import in some of the files, which suggests to me we need to invent at least one test that would have caught this. It also means we have some refactoring ahead of us as it is far too tedious to maintain this many copies of exactly the same import statements etc, but that is for another PR post-v0.1

MechCoder commented:
Fixed and tested the bug in the last commit. I'm not sure on what to rename `verbose_func` as, I'm open to suggestions wrt that.

glouppe commented:
This is a lot better! Thanks for the refactoring :)

Now you will say I am just a p*** in the ***, but the reason I suggested to add a callback API was to implement verbosity as a callback. Do you think it would be possible to do that, instead of wrapping calls into `verbose_func`?

glouppe commented:
(and sorry if my earlier comments were not clear enough...)

MechCoder commented:
haha, I had thought about for printing the verbosity, I need `func` but I don't pass that to the callback. How would you suggest to do that?

glouppe commented:
I would define a `VerboseCallback` object implementing a `__call__` method with the expected signature. Defining the callback as an object would allows you to store some state, e.g. for measuring time between two calls to the callback. In this way, you dont even need to pass `func`.

MechCoder commented:
interesting, I hadn't thought of that before.

glouppe commented:
And you could also define extra arguments to be passed at `__init__`, e.g. like a verbosity level.

betatim commented:
As a user I don't really like the idea of having to create some `VerboseThis` object which I then pass as `callbkack=VerboseThis()` to `xxx_minimise()`. Much rather just `verbose=True` and be done with it. If underneath we implement it via the callback list that is fine with me but for the user it has to be simpler IMHO.

In principle the time between calls to `callbacks` should be a good measure of how long it takes to execute one iteration. Unless callbacks start doing weird things that take a lot of time.

MechCoder commented:
I suppose we set it internally. I am not super-clear about the benefits unless I try it out but it makes me suspect that @glouppe has been reading a lot of java code.

glouppe commented:
The API for user should still be verbose=True. I just dont see the point of
having in parallel callback calls + if-statements for verbosity while
everything could be factored out with the same logic.

Internally `verbose=True` will be translated to
`callback.append(VerboseCallback(...))`
and that is it.

On 5 August 2016 at 18:32, Manoj Kumar <notifications@github.com> wrote:

> I suppose we set it internally. I am not super-clear about the benefits
> unless I try it out but it makes me suspect that @glouppe
> <https://github.com/glouppe> has been reading a lot of java code.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-optimize/scikit-optimize/pull/145#issuecomment-237898100>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAdKS1hK06WP663nV1a0eQp6IDB20Djbks5qc2WLgaJpZM4JSeEM>
> .
>

MechCoder commented:
Yeah, I don't disagree.

On Fri, Aug 5, 2016 at 10:42 AM, Gilles Louppe <notifications@github.com>
wrote:

> The API for user should still be verbose=True. I just dont see the point of
> having in parallel callback calls + if-statements for verbosity while
> everything could be factored out with the same logic.
>
> Internally `verbose=True` will be translated to
> `callback.append(VerboseCallback(...))`
> and that is it.
>
> On 5 August 2016 at 18:32, Manoj Kumar <notifications@github.com> wrote:
>
> > I suppose we set it internally. I am not super-clear about the benefits
> > unless I try it out but it makes me suspect that @glouppe
> > <https://github.com/glouppe> has been reading a lot of java code.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/scikit-optimize/scikit-optimize/pull/
> 145#issuecomment-237898100>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AAdKS1hK06WP663nV1a0eQp6IDB20Djbks5qc2WLgaJpZM4JSeEM>
>
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-optimize/scikit-optimize/pull/145#issuecomment-237915045>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABx9EH6EOhfsT-vJUTqDkbuxT-YBJXrhks5qc3XsgaJpZM4JSeEM>
> .
>



-- 
Manoj,
http://github.com/MechCoder

MechCoder commented:
@glouppe I've thought of a solution but I'm not sure it is as elegant. At present the if statements for verbosity are already filtered out into the common function `verbose_func`. How about we merge this as is and I open a new issue / pull request to discuss the changes?

MechCoder commented:
@betatim @glouppe I added a VeboseCallback object and added it to `gp_opt`. Let me know of your thoughts. There is one minor issue, how do we take into account the time taken to fit the surrogate model since we would not be calling the other callbacks before and after this surrogate model is fit

glouppe commented:
Let us just report the time per iteration, where an iteration comprises evaluating `func` + fitting a model + optimizing the acquisition.

betatim commented:
please `pep8` me :neckbeard:

betatim commented on skopt/tests/test_common.py:
```diff
@@ -61,28 +66,49 @@ def check_minimizer_bounds(result):
+                                n_calls=7, random_state=1,
+                                verbose=verbose, callback=call)
+
+        assert(not result.models)
```
-> `assert(result.models is None)`

glouppe commented on skopt/dummy_opt.py:
```diff
@@ -59,6 +62,14 @@ def dummy_minimize(func, dimensions, n_calls=100,
+        for long optimization runs.
+
+    * `callback` [callable, list of callables, optional]
+        If callable then `callback(res)` is called after each call to func.
```
or simply state that it is called after each iteration?

MechCoder replied:
IMHO, every call to func is more better, iterations is slightly ambiguous.

betatim replied:
func -> `func`

I have no opinion on func vs iteration

glouppe commented on skopt/utils.py:
```diff
@@ -0,0 +1,103 @@
+    return res
+
+
+def check_callback(callback):
```
Let's move all callback logics in a `callbacks` module?

betatim replied:
Maybe a module for collecting all `check_XXX()` functions?. Hopefully in the future we will have more of these `check_blah()` functions to factor out stuff that is done in all the minimisers.

For the moment I'd be ok with keeping it in `utils`.

glouppe commented on skopt/gp_opt.py:
```diff
@@ -246,6 +285,8 @@ def gp_minimize(func, dimensions, base_estimator=None, alpha=10e-10,
     n_model_iter = n_calls - n_total_init_calls
     for i in range(n_model_iter):
+        if verbose:
+            print("Fitting GaussianProcessRegressor no: %d" % (i + 1))
```
This should be removed

glouppe commented on skopt/tree_opt.py:
```diff
@@ -70,15 +85,30 @@ def _tree_minimize(func, dimensions, base_estimator, n_calls,
     n_model_iter = n_calls - n_total_init_calls
     for i in range(n_model_iter):
+        if verbose:
+            print("Fitting model no: %d" % (i + 1))
```
This should be removed

betatim commented:
Should we state somewhere that stuff in `skopt.utils` is for internal use and might change without notice? Do we even want that?

glouppe commented:
I think it is safe to say we will also write some more callbacks, I would prefer having them grouped together in a non-utils module. But anyhow, not very important for the moment.

MechCoder commented:
I have addressed your comments and this is ready for final rounds.

I ran pep8 and all the warnings seem unrelated to my changes.

betatim commented:
> I ran pep8 and all the warnings seem unrelated to my changes.

Looked like there were some spurious/missing blank lines. If they are gone, Mr :neckbeard: is happy :)

betatim commented on skopt/callbacks.py:
```diff
@@ -0,0 +1,93 @@
+    ----------
+    * `n_init` [int, optional]:
+        Number of points provided by the user which are yet to be
+        evaluated. That is where `len(x0)` is `y0` is None.
```
"That is where `len(x0)` is `y0` is None." wat? I think there is a word missing or something?

MechCoder replied:
Should be "This is equal to `len(x0) when `y0` is None"

MechCoder commented:
@betatim Any other comments? If you are happy, I'll wait for @glouppe to merge.

glouppe commented:
Merging, thanks a lot Manoj! and sorry for the back and forth discussions ;)

MechCoder commented:
Yay! One more to go.

betatim commented:
🚀!
